Add N-Dimensional Lookup Table

mj023 opened the pull request:
### What problem do you want to solve?

This PR is supposed to address #989, while also adding a new Class of N-Dimensional Lookup Tables that will replace the old classes for specific dimensional tables.


### Todo

- [ ] Create N-Dim. Lookup Table Object
- [ ] Replace all occurrences of 1-Dim and 2-Dim Lookup Tables

hmgaudecker reviewed (commented):
I could only take a glance, but it looks great! Could you remove the `Nd` / `_nd` from the names? Now that there is just one left, there is no need to mention it.

mj023 commented:
I think I have replaced everything now. I did not make changes to the parsing of YAML files, the old 1d- and 2d-LookupTable datatypes should probably be replaced with the new one, but I was not sure how the YAML structure would need to be changed to incorporate information about the number of dimensions.

 I had some issues with the Typing, do I have to register the string inside `Float[Array, "n_cols n_rows"]` somewhere? Whenever I change it to something new, ruff does not accept it anymore.

I also fixed a presumably wrong value in the  Renten-Altersgrenzen for women in [f5adc58](https://github.com/iza-institute-of-labor-economics/gettsim/pull/998/commits/f5adc58216229db672bc01b2d4e233ee5b7ac2f7).

MImmesberger commented:
> I also fixed a presumably wrong value in the Renten-Altersgrenzen for women in [f5adc58](https://github.com/iza-institute-of-labor-economics/gettsim/pull/998/commits/f5adc58216229db672bc01b2d4e233ee5b7ac2f7).

Yes, that's a typo. Thanks a lot!

hmgaudecker commented on src/ttsim/tt_dag_elements/param_objects.py:
```diff
@@ -97,29 +98,50 @@ class PiecewisePolynomialParam(ParamObject):
-class ConsecutiveInt2dLookupTableParam(ParamObject):
-    """A parameter with its contents read and converted from a YAML file.
+class ConsecutiveIntLookupTableParamValue:
+    """The parameters expected by lookup_table"""
```
```suggestion
    """The `value` for ConsecutiveIntLookupTable."""
```

hmgaudecker commented on src/_gettsim/wohngeld/miete.py:
```diff
@@ -44,34 +40,23 @@ def max_miete_m_lookup_mit_baujahr(
-    subtract_rows = []
-    for baujahr in baujahre:
+    modified_dict = {}
+    for i in range(len(baujahre)):
+        baujahr = baujahre[i]
```
```suggestion
    for i, baujahr in enumerate(baujahre):
```

hmgaudecker commented on src/_gettsim/wohngeld/miete.py:
```diff
@@ -44,34 +40,23 @@ def max_miete_m_lookup_mit_baujahr(
-    subtract_cols = []
-    subtract_rows = []
-    for baujahr in baujahre:
+    modified_dict = {}
```
```suggestion
    lookup_dict = {}
```

Makes clear this has the same structure as the table below.

hmgaudecker commented on src/_gettsim/wohngeld/miete.py:
```diff
@@ -44,34 +40,23 @@ def max_miete_m_lookup_mit_baujahr(
-    full_lookup_table = xnp.stack(values, axis=0)
-    full_lookup_base_to_subtract_cols = xnp.asarray(subtract_cols)
-    full_lookup_base_to_subtract_rows = xnp.asarray(subtract_rows)
+        modified_dict[i] = this_dict
```
```suggestion
            lookup_dict[i] = this_dict
```

hmgaudecker commented on src/_gettsim/wohngeld/miete.py:
```diff
@@ -44,34 +40,23 @@ def max_miete_m_lookup_mit_baujahr(
-        lookup_base_to_subtract_cols=full_lookup_base_to_subtract_cols,
-        lookup_base_to_subtract_rows=full_lookup_base_to_subtract_rows,
+        lookup_table=get_consecutive_int_lookup_table_param_value(
+            raw=modified_dict, n_dims=3, xnp=xnp
```
```suggestion
            raw=lookup_dict, n_dims=3, xnp=xnp
```

hmgaudecker commented on src/_gettsim/wohngeld/miete.py:
```diff
@@ -230,11 +213,7 @@ def miete_m_hh_mit_baujahr(
-        anzahl_personen_hh - max_miete_m_lookup.lookup_base_to_subtract_rows[selected_bin_index],
-        mietstufe - max_miete_m_lookup.lookup_base_to_subtract_cols[selected_bin_index],
-    ]  # fmt: skip
+    max_miete_m = max_miete_m_lookup.lookup_table.lookup(selected_bin_index,anzahl_personen_hh,mietstufe)  # fmt: skip
```
```suggestion
    max_miete_m = max_miete_m_lookup.lookup_table.lookup(baujahr_index,anzahl_personen_hh,mietstufe)  
```
- Rename variable above, too
- We just skipped that because the old line became very ugly otherwise, not an issue anymore!

hmgaudecker commented on src/ttsim/interface_dag_elements/policy_environment.py:
```diff
@@ -162,33 +161,35 @@ def _get_one_param(  # noqa: PLR0911
         )
         return PiecewisePolynomialParam(**cleaned_spec)
     if spec["type"] == "consecutive_int_1d_lookup_table":
```
We should get rid of the 1d / 2d here, too. In particular, we should be able to infer the depth from the nesting level of the dictionary in `cleaned_spec["value"]`

hmgaudecker commented on src/ttsim/tt_dag_elements/param_objects.py:
```diff
@@ -97,29 +98,50 @@ class PiecewisePolynomialParam(ParamObject):
-    reference: str | None = None
+    def lookup(
+        self: ConsecutiveIntLookupTableParamValue, *args: int
+    ) -> Float[Array, "n_rows n_cols"]:
```
I don't quite get the type hint here. I suppose this is still at the level of the individual row (no vectorization to `n_obs`, hence `int` in the arguments)? Then we should also return a scalar `float | int | bool`.

If the annotation is at the column level, we should hav `*args: IntColumn` and return `FloatColumn | IntColumn | BoolColumn`.

hmgaudecker commented on src/ttsim/tt_dag_elements/__init__.py:
```diff
@@ -46,10 +43,10 @@
     "ColumnObject",
-    "ConsecutiveInt1dLookupTableParam",
-    "ConsecutiveInt1dLookupTableParamValue",
     "ConsecutiveInt2dLookupTableParam",
     "ConsecutiveInt2dLookupTableParamValue",
```
```suggestion
```
I am surprised we don't have any complaints by ruff here...

hmgaudecker reviewed (commented):
Very nice! Basically just some cleanup left.

hmgaudecker reviewed (approved):
Perfect, thanks! 

I just renamed a couple of things and changed some type hints.
